channelserver/src/server.rs

```diff
@@ -0,0 +1,257 @@
+                let msg_len = message.len();
+                if max_data > 0 && (party.data_exchanged > max_data || msg_len > max_data) {
+                    slog_info!(
+                        self.log.log,
+                        "Too much data sent through {}, closing",
+                        channel
+                    );
+                    return Err(perror::HandlerErrorKind::XSDataErr.into());
+                }
+                party.data_exchanged += msg_len;
+                let msg_count = u8::from(self.settings.borrow().max_exchanges);
+                party.msg_count += 1;
+                if msg_count > 0 && party.msg_count > msg_count {
+                    slog_info!(
+                        self.log.log,
+                        "Too many messages through {}, closing",
+                        channel
+                    );
+                    return Err(perror::HandlerErrorKind::XSMessageErr.into());
+                }
+                if party.id != skip_id {
+                    if let Some(addr) = self.sessions.get(&party.id) {
+                        addr.do_send(TextMessage(message.to_owned())).unwrap_or(());
+                    }
+                } else {
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Kill a channel and terminate all participants.
+    ///
+    /// This sends a ^D message to each participant, which forces the connection closed.
+    fn shutdown(&mut self, channel: &Uuid) {
+        if let Some(participants) = self.channels.get_mut(channel) {
+            for (id, info) in participants {
+                if let Some(addr) = self.sessions.get(&id) {
+                    // send a control message to force close
+                    addr.do_send(TextMessage("\x04".to_owned())).unwrap_or(());
+                }
+                self.sessions.remove(&id);
+            }
+        }
+    }
+}
+
+/// Make actor from `ChannelServer`
+impl Actor for ChannelServer {
+    /// We are going to use simple Context, we just need ability to communicate
+    /// with other actors.
+    type Context = Context<Self>;
+}
+
+/// Handler for Connect message.
+///
+/// Register new session and assign unique id to this session
+impl Handler<Connect> for ChannelServer {
+    type Result = usize;
+
+    fn handle(&mut self, msg: Connect, ctx: &mut Context<Self>) -> Self::Result {
+        let id = self.rng.borrow_mut().gen::<usize>();
+        let new_chan = Channel {
+            // register session with random id
+            id: id.clone(),
+            started: Instant::now(),
+            msg_count: 0,
+            data_exchanged: 0,
+        };
+        {
+            self.sessions.insert(new_chan.id, msg.addr.clone());
+        }
+        slog_debug!(
+            self.log.log,
+            "New connection to {} for  {} : [{}]",
+            &self.me,
+            &msg.channel.simple(),
+            &new_chan.id
+        );
+
+        // auto join session to Main channel
```

Comment:
I don't know what "Main channel" is, is this comment out of date?

Reply:
Yes, again, part of the original demo.